### ISSUE TITLE
distance to edge for sand/hard-bottom RAM issues

### ISSUE THREAD

**Issue by dwwolfson** [dwwolfson/red_snapper#3]
Hey all,
I've been trying to figure out the most efficient way we can calculate the distance to edge value for every pixel cell like we've been discussing. Trouble is, with 2.8M pixels and a fairly complex line edge to measure to, my computer can't handle it. I converted the edge of sand and hard-bottom to a line, and each pixel from the raster to a point. When I try to calculate a distance matrix I get the error: "cannot allocate a vector of size 15.6Gb". 

I'm trying to base the current approach on this SO post:
https://gis.stackexchange.com/questions/233443/finding-distance-between-raster-pixels-and-line-features-in-r

I'm no expert in this, so feel free to explore other directions. I'm guess there is a better way to try out there. I was also thinking maybe we could try overlaying the line features onto the raster and if there was an intersection, give that pixel a value of 1, otherwise 0 or NA. Then I think there are functions to measure from each pixel to the closest next pixel that has a certain value or is NA.

**Comment by dwwolfson** [comment]
Tried a few other approaches:
If i consolidate the spatial lines object with `gUnion(obj, obj)` prior to using it with `gDistance`, it will run, but it never finished. I ended the operation after 10 hours.

I also tried converting the seabed raster to SpatialPoints and then to an sf object so that I could use `st_distance`:
```
seabed<-st_read(here("data/Seabed Maps/Shapefile/NF_19_06_Segmentation.shp"))
seabed<-seabed %>% 
  mutate(binary_hab=ifelse(Class_name%in%c("High", "Low", "Medium"), "hard-bottom", "sand"))

seabed<-seabed %>% 
  group_by(binary_hab) %>% 
  summarise(geometry=st_union(geometry))
sea_lines<-st_cast(seabed, "MULTILINESTRING")
sea_iso<-st_cast(sea_lines, "LINESTRING")

ras<-raster(here("data/Seabed Maps/geotiff/", "ChickenRock_Classification.tif"))

p<-as(ras, "SpatialPoints")
p_sf<-st_as_sf(p)

closest <- list()
for(i in seq_len(nrow(p_sf))){
  closest[[i]] <- sea_iso[which.min(
    st_distance(sea_iso, p_sf[i,])),]
}
```

This also didn't finish. It's been almost 24 hours so I'm going to stop it. I'll try again with a few points at a time and verify that it's actually working, and then maybe also add in a counter that will print to the console so I know the progress. There is a lot to do with >2M pixels/points!

**Comment by dwwolfson** [comment]
Bad news, good news, and bad news.
The latest attempt with `st_distance` was working, but only had saved 116K points when I stopped it. Fortunately, I realized that I was saving each result as it's own sf object, which isn't necessary. After converting to just
```
closest[[i]]<-min(st_distance(sea_iso, p_sf[i,]))
```
it looks like it cuts the time in half. Unfortunately, that still means that it will take ~5.5 days to do the whole thing. Maybe I can try doing this on the remote computer, but I'm worried then it will take even longer since those computers are pretty bare bones.

Maybe a compromise is to use this for our actual snapper points so that we can include the covariate in the model, then get a computer chugging away on doing all pixels for the entire raster layer. Unless.......there is a significantly better/faster way.
